src/server.rs

```diff
@@ -335,80 +329,135 @@ fn collect_changes(
+                        .insert(archetype_entity.entity(), change_tick.this_run());
+                } else {
+                    update_message.register_entity();
+                    update_message.end_entity_data()?;
+                }
+
+                init_message.end_entity_data()?;
             }
         }
     }
 
-    for message in &mut *messages {
-        message.end_array()?;
+    for (init_message, _) in messages.iter_mut() {
+        init_message.end_array()?;
+    }
+
+    Ok(())
+}
+
+/// Collects the component if it has been changed.
+///
+/// If the component has been changed in this tick, it will be collected into init buffer.
+/// Otherwise if the component has been changed since the last entity tick for a client - it will be collected into update message.
+fn collect_component_change(
+    messages: &mut ReplicationMessages,
+    entity: Entity,
+    ticks: ComponentTicks,
+    change_tick: &SystemChangeTick,
+    replication_info: &ReplicationInfo,
+    replication_id: ReplicationId,
+    component: Ptr,
+) -> bincode::Result<()> {
+    for (init_message, update_message, client_info) in messages.iter_mut_with_info() {
+        if ticks.is_added(change_tick.last_run(), change_tick.this_run()) {
+            init_message.write_component(replication_info, replication_id, component)?;
+        } else {
+            let tick = *client_info
+                .ticks
+                .get(&entity)
+                .expect("entity should present after adding component");
+            if ticks.is_changed(tick, change_tick.this_run()) {
+                update_message.write_component(replication_info, replication_id, component)?;
+            }
+        }
     }
 
     Ok(())
 }
 
-/// Collect component removals into messages based on last acknowledged tick.
+/// Collects component removals from this tick into init messages.
 fn collect_removals(
-    messages: &mut [ReplicationMessage],
-    removal_trackers: &Query<(Entity, &RemovalTracker)>,
-    system_tick: Tick,
+    messages: &mut ReplicationMessages,
+    remove_events: &RemovedComponentEvents,
+    tick: Tick,
+    replication_rules: &ReplicationRules,
 ) -> bincode::Result<()> {
-    for message in &mut *messages {
+    for (message, _) in messages.iter_mut() {
         message.start_array();
     }
 
-    for (entity, removal_tracker) in removal_trackers {
-        for message in &mut *messages {
+    // PERF: Unfortunately, removed components are grouped by type, not by entity.
+    // This is why we need an intermediate container. But in practice users rarely
+    // remove a lot of components in the same tick, so it's probably fine.
+    let mut removals: EntityHashMap<_, Vec<_>> = Default::default();
+    for (&component_id, &replication_id) in replication_rules.get_ids() {
+        for entity in remove_events
+            .get(component_id)
+            .into_iter()
+            .flat_map(|removed| removed.iter_current_update_events().cloned())
```

Comment:
Good catch, thank you! But since that's what we used before (I just moved it here), I will investigate it after this PR. Opened issue: https://github.com/lifescapegame/bevy_replicon/issues/126.